pkgs/native_assets_builder/lib/src/build_runner/build_runner.dart

```diff
@@ -47,39 +50,77 @@ class NativeAssetsBuildRunner {
-      buildPlan = packagesWithNativeAssets;
-      packageGraph = PackageGraph({
-        for (final p in packagesWithNativeAssets) p.name: [],
-      });
-    } else {
-      final planner = await NativeAssetsBuildPlanner.fromRootPackageRoot(
-        rootPackageRoot: packageLayout.rootPackageRoot,
-        packagesWithNativeAssets: packagesWithNativeAssets,
-        dartExecutable: Uri.file(Platform.resolvedExecutable),
-        logger: logger,
-      );
-      final (plan, planSuccess) = planner.plan(
+  }) async =>
+      _run(
+        step: PipelineStep.build,
+        linkModePreference: linkModePreference,
+        target: target,
+        workingDirectory: workingDirectory,
+        buildMode: buildMode,
+        cCompilerConfig: cCompilerConfig,
+        targetIOSSdk: targetIOSSdk,
+        targetAndroidNdkApi: targetAndroidNdkApi,
+        includeParentEnvironment: includeParentEnvironment,
+        packageLayout: packageLayout,
         runPackageName: runPackageName,
+        supportedAssetTypes: supportedAssetTypes,
       );
-      if (!planSuccess) {
-        return _BuildResultImpl(
-          assets: [],
-          dependencies: [],
-          success: false,
-        );
-      }
-      buildPlan = plan;
-      packageGraph = planner.packageGraph;
+
+  Future<LinkResult> link({
+    required Target target,
+    required Uri workingDirectory,
+    required BuildModeImpl buildMode,
+    CCompilerConfigImpl? cCompilerConfig,
+    IOSSdkImpl? targetIOSSdk,
+    int? targetAndroidNdkApi,
+    required bool includeParentEnvironment,
+    PackageLayout? packageLayout,
+    Uri? resourceIdentifiers,
+    String? runPackageName,
+    Iterable<String>? supportedAssetTypes,
+    required BuildResult buildResult,
+  }) async =>
+      _run(
+        step: PipelineStep.link,
+        linkModePreference: LinkModePreferenceImpl.dynamic,
+        target: target,
+        workingDirectory: workingDirectory,
+        buildMode: buildMode,
+        cCompilerConfig: cCompilerConfig,
+        targetIOSSdk: targetIOSSdk,
+        targetAndroidNdkApi: targetAndroidNdkApi,
+        includeParentEnvironment: includeParentEnvironment,
+        packageLayout: packageLayout,
+        runPackageName: runPackageName,
+        resourceIdentifiers: resourceIdentifiers,
+        supportedAssetTypes: supportedAssetTypes,
+        previousBuildResult: buildResult,
+      ).then(LinkResult.new);
+
+  Future<BuildResultImpl> _run({
+    required PipelineStep step,
+    required LinkModePreferenceImpl linkModePreference,
+    required Target target,
+    required Uri workingDirectory,
+    required BuildModeImpl buildMode,
+    CCompilerConfigImpl? cCompilerConfig,
+    IOSSdkImpl? targetIOSSdk,
+    int? targetAndroidNdkApi,
+    required bool includeParentEnvironment,
+    PackageLayout? packageLayout,
+    Uri? resourceIdentifiers,
+    String? runPackageName,
+    Iterable<String>? supportedAssetTypes,
+    BuildResult? previousBuildResult,
```

Comment:
just buildResult?

and maybe add an assert that it is null if pipilinestep==build